File: src/optical_element/MATLAB/g_paraboid_focus.m
classdef  g_paraboid_focus< g_cyliner_parabolic_focus%追迹phase转换成角度
  properties 

  end
methods
function obj = g_paraboid_focus( beam_in,ds,di,chi,theta,surface,r2,grating)   
    obj=obj@g_cyliner_parabolic_focus( beam_in,ds,di,chi,theta,surface,r2,grating) ;%输入参数给父类
end
end

methods%在光学镜坐标系中，计算光束和光学元器件的交点
    %[X2,Y2,Z2]是交点，默认情况是平面镜的交叉点计算方法
    function [X2,Y2,Z2,T]=intersection(obj)
        L1=obj.L1;M1=obj.M1;N1=obj.N1;
        X1=obj.X1;Y1=obj.Y1;Z1=obj.Z1;
        
        A = M1.^2+L1.^2;
        B = 2*(N1.*obj.p+M1.*Y1+L1.*X1);
        C = X1.^2+Y1.^2+2*obj.p.*Z1;
        T = (-B+(B.^2-4*A.*C).^0.5)./(2*A);
        X2 = X1+T.*L1;                                                              %Nx=0
        Y2 = Y1+T.*M1; 
        Z2 = Z1+T.*N1;        
        obj.T=T;
    end
end



methods%在光学平面坐标系中，平面镜的法线方向
    function [Nx,Ny,Nz]=normal(obj)
        Y2=obj.Y2;%Z2=obj.Z2;
        X2=obj.X2;
        Nx = -X2./(X2.^2+Y2.^2+obj.p^2).^0.5;
        Ny = -Y2./(X2.^2+Y2.^2+obj.p^2).^0.5;  
        Nz = -obj.p./(X2.^2+Y2.^2+obj.p^2).^0.5; 
        obj.Nx=Nx;
        obj.Ny=Ny;
        obj.Nz=Nz;
    end
end

end

File: src/optical_element/g_paraboid_focus.cpp
#include "g_paraboid_focus.h"

using namespace Furion_NS;

G_Paraboid_Focus::G_Paraboid_Focus(G_Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, double r2, Grating* grating)
    : G_Cyliner_Parabolic_Focus(beam_in, ds, di, chi, theta, surface, r2, grating)
{
    cout << "G_Paraboid_Focus的初始化" << endl;
    run(beam_in, ds, di, chi, theta, surface, r2, grating);
}

G_Paraboid_Focus::~G_Paraboid_Focus()
{
    //delete[] T, Nx, Ny, Nz;
}

void G_Paraboid_Focus::run(G_Beam* beam_in, double ds, double di, double chi, double theta, No_Surfe* surface, double r2, Grating* grating)
{
    cout << "G_Paraboid_Focus的run" << endl;
    reflect(beam_in, ds, di, chi, theta);
}

void G_Paraboid_Focus::intersection(std::vector<double>& T)
{
    int n = this->n;

    //double *A= new double[Furion::n];
    //double *B= new double[Furion::n];
    //double *C= new double[Furion::n];
    //std::vector<double> A(n);
    //std::vector<double> B(n);
    //std::vector<double> C(n);
    double A, B, C;

    for (int i = 0; i < n; i++) 
    {
        A = this->M1[i] * this->M1[i] + this->L1[i] * this->L1[i];
        B = 2 * (this->p * this->N1[i] + this->M1[i] * this->Y1[i] + this->L1[i] * this->X1[i]);
        C = this->X1[i] * this->X1[i] + this->Y1[i] * this->Y1[i] + 2 * this->p * this->Z1[i];
        T[i] = (-B + sqrt(B * B - 4 * A * C)) / (2 * A);
        this->T[i] = T[i];

        this->X2[i] = this->X1[i] + T[i]*this->L1[i];
        this->Y2[i] = this->Y1[i] + T[i]*this->M1[i];
        this->Z2[i] = this->Z1[i] + T[i]*this->N1[i];
    } 
    cout << "G_Paraboid_Focus的intersection" << endl;
    
    //destory_1d(A);
    //destory_1d(B);
    //destory_1d(C);
}

void G_Paraboid_Focus::normal(std::vector<double>& Nx, std::vector<double>& Ny, std::vector<double>& Nz)
{
    int n = this->n;

    double data;
    double p2 = this->p * this->p;

    for (int i = 0; i < n; i++) 
    {
        data = sqrt(this->X2[i]*this->X2[i] + this->Y2[i] * this->Y2[i] + p2);
        Nx[i] = -(this->X2[i]) / data;
        Ny[i] = -(this->Y2[i]) / data;
        Nz[i] = -this->p / data;

        this->Nx[i] = Nx[i];
        this->Ny[i] = Ny[i];
        this->Nz[i] = Nz[i];
    }    
    cout << "G_Paraboid_Focus的normal" << endl;
}